mono/sgen/sgen-gray.c

```diff
@@ -146,14 +161,33 @@ sgen_gray_object_dequeue (SgenGrayQueue *queue)
 #endif
 
 	if (G_UNLIKELY (queue->cursor < GRAY_FIRST_CURSOR_POSITION (queue->first))) {
-		GrayQueueSection *section = queue->first;
+		GrayQueueSection *section;
+		gint32 old_num_sections;
+
+		old_num_sections = InterlockedDecrement (&queue->num_sections);
+
+		if (old_num_sections <= 0) {
+			mono_os_mutex_lock (&queue->steal_mutex);
```

Comment:
Wait, so if there are more sections left we allow multiple workers to steal lock-free?